src/server/socksServer.ts

```diff
@@ -0,0 +1,312 @@
+          else if (this._addressType === SOCKS_ATYP.NAME)
+            this._phase = ConnectionPhases.REQ_DSTADDR_VARLEN;
+          i++;
+          break;
+
+        case ConnectionPhases.REQ_DSTADDR: {
+          assert(this._dstAddr);
+          const left = this._dstAddr.length - this._dstAddrp;
+          const chunkLeft = chunk.length - i;
+          const minLen = (left < chunkLeft ? left : chunkLeft);
+          chunk.copy(this._dstAddr, this._dstAddrp, i, i + minLen);
+          this._dstAddrp += minLen;
+          i += minLen;
+          if (this._dstAddrp === this._dstAddr.length)
+            this._phase = ConnectionPhases.REQ_DSTPORT;
+          break;
+        }
+
+        case ConnectionPhases.REQ_DSTADDR_VARLEN:
+          this._dstAddr = Buffer.alloc(chunk[i]);
+          this._phase = ConnectionPhases.REQ_DSTADDR;
+          i++;
+          break;
+
+        case ConnectionPhases.REQ_DSTPORT:
+          assert(this._dstAddr);
+          if (this._dstPort === undefined) {
+            this._dstPort = chunk[i];
+          } else {
+            this._dstPort <<= 8;
+            this._dstPort += chunk[i];
+            i++;
+
+            this._socket.removeListener('data', this._onData);
+            if (i < chunk.length)
+              this._socket.unshift(chunk.slice(i));
+
+            if (this._addressType === SOCKS_ATYP.IPv4) {
+              this._info.dstAddr = [...this._dstAddr].join('.');
+            } else if (this._addressType === SOCKS_ATYP.IPv6) {
+              let ipv6str = '';
+              const addr = this._dstAddr;
+              for (let b = 0; b < 16; ++b) {
+                if (b % 2 === 0 && b > 0)
+                  ipv6str += ':';
+                ipv6str += (addr[b] < 16 ? '0' : '') + addr[b].toString(16);
+              }
+              this._info.dstAddr = ipv6str;
+            } else {
+              this._info.dstAddr = this._dstAddr.toString();
+            }
+            this._info.dstPort = this._dstPort;
+            this._phase++;
+            this._readyResolve();
+            return;
+          }
+          i++;
+          break;
+        default:
+          assert(false);
+      }
+    }
+  }
+
+  private _authWithoutPassword(socket: net.Socket) {
+    socket.write(Buffer.from([0x05, 0x00]));
+  }
+
+  async ready(): Promise<{ info: SocksConnectionInfo, forward: () => void, intercept: (parent: SdkObject) => SocksInterceptedSocketHandler }> {
+    await this._ready;
+    return {
+      info: this._info,
+      forward: () => {
+        const dstSocket = new net.Socket();
+        this._socket.on('close', () => dstSocket.end());
+        this._socket.on('end', () => dstSocket.end());
+        dstSocket.setKeepAlive(false);
+        dstSocket.on('error', (err: NodeJS.ErrnoException) => writeSocksSocketError(this._socket, String(err)));
+        dstSocket.on('connect', () => {
+          this._socket.write(BUF_REP_INTR_SUCCESS);
+          this._socket.pipe(dstSocket).pipe(this._socket);
```

Comment:
Interesting! So we are supposed to pipe even the "pass-through" sockets? I was under impression we can just reply with the "BND == DST" and let the client connect on its own, but maybe I misread that.